src/fat/block_translator.rs

```diff
@@ -0,0 +1,383 @@
+                && b1.stored_block == b2.stored_block)));
+
+        let real_block = block_in_vbs / BlockIndex::from(self.vbs_per_hbs);
+        #[allow(clippy::cast_possible_truncation)]
+        let offset = (block_in_vbs % BlockIndex::from(self.vbs_per_hbs)) as usize;
+
+        let buffer = match BUFS {
+            1 => {
+                let buffer = &mut self.buffers[0];
+
+                if buffer.stored_block == real_block
+                    && buffer.status != BlockTranslatorStatus::Unknown
+                {
+                    return Ok((buffer, offset));
+                }
+
+                buffer
+            }
+            2 => {
+                // check newest buffer
+                if self.buffers[self.next].stored_block == real_block
+                    && self.buffers[self.next].status != BlockTranslatorStatus::Unknown
+                {
+                    return Ok((&mut self.buffers[self.next], offset));
+                }
+
+                // switch to older buffer
+                self.next ^= 1;
+
+                let buffer = &mut self.buffers[self.next];
+
+                // check older buffer
+                if buffer.status != BlockTranslatorStatus::Unknown
+                    && buffer.stored_block == real_block
+                {
+                    return Ok((buffer, offset));
+                }
+
+                buffer
+            }
+            _ => {
+                if self.next == usize::MAX {
+                    // reset all ages because an overflow would happen otherwise
+                    let mut ages = [(0, 0); BUFS];
+                    for (num, buf) in self.buffers.iter().enumerate() {
+                        ages[num] = (num, buf.last_used);
+                    }
+                    ages.sort_by_key(|&(_, age)| age);
+                    for (new_age, (num, _last_used)) in ages.into_iter().enumerate() {
+                        self.buffers[num].last_used = new_age;
+                    }
+                    self.next = BUFS;
+                }
+
+                let mut oldest: Option<&'b mut Buffer<'a, BUF_SIZE>> = None;
+
+                // check if the block is already buffered
+                for buffer in self.buffers.iter_mut() {
+                    if buffer.stored_block == real_block
+                        && buffer.status != BlockTranslatorStatus::Unknown
+                    {
+                        // update last_used, unless it already was the last used
+                        if buffer.last_used != self.next - 1 {
+                            buffer.last_used = self.next;
+                            self.next += 1;
+                        }
+
+                        return Ok((buffer, offset));
+                    }
+                    if oldest.is_none() || buffer.last_used < oldest.as_ref().unwrap().last_used {
+                        oldest = Some(buffer);
+                    }
+                }
```

Comment:
I see some memoization potential here (since each buffer's location doesn't change, perhaps we could move all the last usage data into a separate struct field?)